pkg/vault/contracts/WrappedBalancerPoolToken.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { ERC20Permit } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
+
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IWrappedBalancerPoolToken } from "@balancer-labs/v3-interfaces/contracts/vault/IWrappedBalancerPoolToken.sol";
+
+/**
+ * @notice ERC20 wrapper for Balancer Pool Token (BPT), allowing users to deposit BPT
+ * and receive 1:1 wrapped tokens, or burn wrapped tokens to redeem the original BPT.
+ * Minting and burning are only allowed when the Vault is locked.
+ */
+contract WrappedBalancerPoolToken is IWrappedBalancerPoolToken, ERC20, ERC20Permit {
+    using SafeERC20 for *;
+
+    IERC20 public immutable bpt;
+    IVault public immutable vault;
+
+    constructor(
+        IVault vault_,
+        IERC20 bpt_,
+        string memory name,
+        string memory symbol
+    ) ERC20(name, symbol) ERC20Permit(name) {
+        vault = vault_;
+        bpt = bpt_;
+    }
+
+    modifier onlyIfVaultLocked() {
+        if (vault.isUnlocked()) {
+            revert VaultIsUnlocked();
+        }
+        _;
+    }
+
+    /// @inheritdoc IWrappedBalancerPoolToken
+    function mint(uint256 amount) public onlyIfVaultLocked {
+        bpt.safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
Note: BPTs can't be _completely_ trusted.

It'd be a very weird edge case, since this contract is just to wrap certain BPTs that work in lending protocols, and as such they should just be standard implementations. I guess if there's something malicious about the transfer function the wrapper would just not make any sense either right? In other words, maybe this is just fine as is, and there's no need to add any extra protections.